File: module-git/src/main/java/com/timecat/module/git/sgit/repo/tasks/repo/CommitChangesTask.java
package com.timecat.module.git.sgit.repo.tasks.repo;

import android.content.Context;
import android.widget.Toast;

import com.timecat.extend.arms.BaseApplication;
import com.timecat.module.git.R;
import com.timecat.module.git.android.utils.Profile;
import com.timecat.module.git.sgit.database.models.Repo;
import com.timecat.module.git.sgit.exception.StopTaskException;

import org.eclipse.jgit.api.CommitCommand;
import org.eclipse.jgit.api.errors.ConcurrentRefUpdateException;
import org.eclipse.jgit.api.errors.GitAPIException;
import org.eclipse.jgit.api.errors.NoHeadException;
import org.eclipse.jgit.api.errors.NoMessageException;
import org.eclipse.jgit.api.errors.UnmergedPathsException;
import org.eclipse.jgit.api.errors.WrongRepositoryStateException;
import org.eclipse.jgit.lib.StoredConfig;

public class CommitChangesTask extends RepoOpTask {

    private AsyncTaskPostCallback mCallback;
    private String mCommitMsg;
    private String mAuthorName;
    private String mAuthorEmail;
    private boolean mIsAmend;
    private boolean mStageAll;

    public CommitChangesTask(Repo repo, String commitMsg, boolean isAmend,
                             boolean stageAll, String authorName, String authorEmail,
                             AsyncTaskPostCallback callback) {
        super(repo);
        mCallback = callback;
        mCommitMsg = commitMsg;
        mIsAmend = isAmend;
        mStageAll = stageAll;
        mAuthorName = authorName;
        mAuthorEmail = authorEmail;
        setSuccessMsg(R.string.git_success_commit);
    }

    @Override
    protected Boolean doInBackground(Void... params) {
        return commit();
    }

    protected void onPostExecute(Boolean isSuccess) {
        super.onPostExecute(isSuccess);
        if (mCallback != null) {
            mCallback.onPostExecute(isSuccess);
        }
    }

    public boolean commit() {
        try {
            commit(mRepo, mStageAll, mIsAmend, mCommitMsg, mAuthorName, mAuthorEmail);
        } catch (StopTaskException e) {
            return false;
        } catch (GitAPIException e) {
            setException(e);
            return false;
        } catch (Throwable e) {
            setException(e);
            return false;
        }
        mRepo.updateLatestCommitInfo();
        return true;
    }

    public static void commit(Repo repo, boolean stageAll, boolean isAmend,
                              String msg, String authorName, String authorEmail) throws Exception, NoHeadException, NoMessageException,
            UnmergedPathsException, ConcurrentRefUpdateException,
            WrongRepositoryStateException, GitAPIException, StopTaskException {
        Context context = BaseApplication.getContext();
        StoredConfig config = repo.getGit().getRepository().getConfig();
        String committerEmail = config.getString("user", null, "email");
        String committerName = config.getString("user", null, "name");

        if (committerName == null || committerName.equals("")) {
            committerName = Profile.getUsername(context);
        }
        if (committerEmail == null || committerEmail.equals("")) {
            committerEmail = Profile.getEmail(context);
        }
        if (committerName.isEmpty() || committerEmail.isEmpty()) {
            Toast.makeText(context, "请在设置中设置用户名和邮箱", Toast.LENGTH_LONG).show();
            return;
        }
        if (msg.isEmpty()) {
            Toast.makeText(context, "提交信息未填写", Toast.LENGTH_LONG).show();
            return;
        }
        CommitCommand cc = repo.getGit().commit()
                .setCommitter(committerName, committerEmail).setAll(stageAll)
                .setAmend(isAmend).setMessage(msg);
        if (authorName != null && authorEmail != null) {
            cc.setAuthor(authorName, authorEmail);
        }
        cc.call();
        repo.updateLatestCommitInfo();
    }
}

File: module-git/src/main/java/com/timecat/module/git/tasks/CommitChangesTask.kt
package com.timecat.module.git.tasks

import android.content.Context
import android.widget.Toast
import com.timecat.extend.arms.BaseApplication
import com.timecat.module.git.R
import com.timecat.module.git.utils.Profile
import com.timecat.module.git.sgit.database.models.Repo
import com.timecat.module.git.utils.StopTaskException
import org.eclipse.jgit.api.errors.*

class CommitChangesTask(
    repo: Repo, private val mCommitMsg: String, private val mIsAmend: Boolean,
    private val mStageAll: Boolean, private val mAuthorName: String, private val mAuthorEmail: String,
    private val mCallback: AsyncTaskPostCallback?
) : RepoOpTask(repo) {
    init {
        setSuccessMsg(R.string.git_success_commit)
    }

    override fun doInBackground(vararg params: Void): Boolean {
        return commit()
    }

    override fun onPostExecute(isSuccess: Boolean) {
        super.onPostExecute(isSuccess)
        mCallback?.onPostExecute(isSuccess)
    }

    fun commit(): Boolean {
        try {
            commit(mRepo, mStageAll, mIsAmend, mCommitMsg, mAuthorName, mAuthorEmail)
        } catch (e: StopTaskException) {
            return false
        } catch (e: GitAPIException) {
            setException(e)
            return false
        } catch (e: Throwable) {
            setException(e)
            return false
        }
        mRepo.updateLatestCommitInfo()
        return true
    }

    companion object {
        @Throws(
            Exception::class,
            NoHeadException::class,
            NoMessageException::class,
            UnmergedPathsException::class,
            ConcurrentRefUpdateException::class,
            WrongRepositoryStateException::class,
            GitAPIException::class,
            StopTaskException::class
        )
        fun commit(
            repo: Repo,
            stageAll: Boolean,
            isAmend: Boolean,
            msg: String,
            authorName: String?,
            authorEmail: String?
        ) {
            val context: Context = BaseApplication.getContext()
            val git = repo.getGit() ?: return
            val config = git.repository.config
            var committerEmail = config.getString("user", null, "email")
            var committerName = config.getString("user", null, "name")
            if (committerName == null || committerName == "") {
                committerName = Profile.getUsername(context)
            }
            if (committerEmail == null || committerEmail == "") {
                committerEmail = Profile.getEmail(context)
            }
            if (committerName!!.isEmpty() || committerEmail!!.isEmpty()) {
                Toast.makeText(context, "请在设置中设置用户名和邮箱", Toast.LENGTH_LONG).show()
                return
            }
            if (msg.isEmpty()) {
                Toast.makeText(context, "提交信息未填写", Toast.LENGTH_LONG).show()
                return
            }
            val cc = git.commit()
                .setCommitter(committerName, committerEmail).setAll(stageAll)
                .setAmend(isAmend).setMessage(msg)
            if (authorName != null && authorEmail != null) {
                cc.setAuthor(authorName, authorEmail)
            }
            cc.call()
            repo.updateLatestCommitInfo()
        }
    }
}